File: beatdetect-model/beatdetect_model/model/postprocessing/inference.py
from dataclasses import dataclass, field
from typing import Self

import numpy as np

from .positions import frames_per_position, pos_per_beat


def emission_lp(is_downbeat, is_beat, observation):
    beat_prob = observation[0]
    downbeat_prob = observation[1]
    no_beat_prob = observation[2]
    if is_downbeat:
        return np.log(downbeat_prob + 1e-9)
    elif is_beat:
        return np.log(beat_prob + 1e-9)
    else:
        return np.log(no_beat_prob + 1e-9)


@dataclass(order=True)
class Node:
    lp: float
    state: tuple = field(compare=False)
    prev: Self = field(compare=False, default=None)

    # need to be rounded when use for indexing
    frame: float = field(compare=False, default=0)


def backtrack(node: Node, fps, periods):
    path = []
    while node is not None:
        path.append(node)
        node = node.prev
    path.reverse()

    beats = []
    t = 0
    for node in path:
        period_idx, _, phase = node.state
        period = periods[period_idx]
        if phase % pos_per_beat(period, fps) == 0:
            beat_num = phase // pos_per_beat(period, fps)
            beats.append([t, beat_num + 1])
        t += frames_per_position(period, fps) / fps
    return np.asarray(beats)


def default_beam_width(T, c=450000, w_min=64, w_max=1024):
    return np.clip(round(c / T), w_min, w_max)


def beam_search(config, nn_output, beam_width=None):
    T = nn_output.shape[1]
    fps = config.spectrogram.fps

    if beam_width is None:
        beam_width = default_beam_width(T)

    periods = sorted(
        [(60 / tempo) * config.spectrogram.fps for tempo in config.post.tempo_bins]
    )

    # wrapper for emission_lp
    def calc_emit_lp(state, observation):
        period_idx, _ts, phase = state
        period = periods[period_idx]
        is_downbeat = phase == 0
        is_beat = phase % pos_per_beat(period, fps) == 0
        return emission_lp(is_downbeat, is_beat, observation)

    transitions_file = np.load(config.paths.transitions)
    row_ptrs = transitions_file["row_ptrs"]
    transition_lp = transitions_file["lp"]
    states = transitions_file["states"]
    states_to_idx = transitions_file["states_to_idx"]
    possible_next_states = transitions_file["possible_next_states"]
    init_dist_file = np.load(config.paths.init_dist)
    init_period_dist = init_dist_file["period"]

    start_states_probs = np.array(
        [
            init_period_dist[state[0]] + calc_emit_lp(state, nn_output[:, 0])
            for state in states
        ]
    )  # not normalized, which is probably fine
    start_states_idxs = np.argsort(start_states_probs)[-beam_width:]
    beam = [
        Node(start_states_probs[state_idx], states[state_idx])
        for state_idx in start_states_idxs
    ]
    best_node, best_lp = None, -np.inf

    while beam:
        node = beam.pop()
        period_idx = node.state[0]
        period = periods[period_idx]
        # Frame of next position
        next_frame = node.frame + frames_per_position(period, fps)

        if round(next_frame) >= T:
            if node.lp > best_lp:
                best_node = node
                best_lp = node.lp
            continue

        next_observation = nn_output[:, round(next_frame)]

        # Expand transitions
        state_idx = states_to_idx[*node.state]
        start, end = row_ptrs[state_idx], row_ptrs[state_idx + 1]
        for p in range(start, end):
            next_state_idx = possible_next_states[p]

            next_state = states[next_state_idx]
            emit_lp = calc_emit_lp(next_state, next_observation)

            next_lp = node.lp + transition_lp[p] + emit_lp
            beam.append(Node(next_lp, next_state, node, next_frame))

        beam.sort(key=lambda n: n.lp)
        beam = beam[-beam_width:]
        # # heapq only faster when avg children > 100-200
        # beam = heapq.nlargest(beam_width, beam)[::-1]

    beats = backtrack(best_node, fps, periods)
    return beats, best_lp

File: beatdetect-web/src/lib/utils/inference.ts
import type { Config } from '$lib/types/config';
import JSZip from 'jszip';
import { load, type NpyArray } from 'npyjs';

import { fetchBinaryCached, getModelURL } from '$lib/utils/model-registry';

// Type definitions
type State = [number, number, number]; // period idx, ts idx, phase
type Observation = [number, number, number]; // beat prob, down beat prob, no beat prob

interface Arrays {
  rowPtrs: Int32Array; // [numStates + 1]
  possibleNextStatesFlat: Int32Array; // flat transition target list [n_transitions]
  transLP: Float32Array; // transition log probs [n_transitions]
  statesFlat: Int32Array; // flatten [numStates * 3]
  statesToIdxFlat: Int32Array; // flatten 3D array
  statesToIdxShape: [number, number, number]; // [P, TS, maxPhase]
  initPeriodDist: Float32Array; // [numPeriods]
}

async function loadNPZ(buffer: ArrayBuffer | Uint8Array) {
  const zip = await JSZip.loadAsync(buffer);
  const result: Record<string, NpyArray> = {};

  for (const [filename, file] of Object.entries(zip.files)) {
    if (!filename.endsWith('.npy')) continue;

    const npyBuffer = await file.async('arraybuffer');
    const arr = await load(npyBuffer);
    result[filename.replace(/\.npy$/, '')] = arr;
  }

  return result;
}

export async function loadArrays(): Promise<Arrays> {
  const result: Partial<Arrays> = {};

  const transUrl = await getModelURL('transitions');
  const initUrl = await getModelURL('init_dist');

  const transBuf = await fetchBinaryCached(transUrl);
  const initBuf = await fetchBinaryCached(initUrl);

  for (const buffer of [transBuf, initBuf]) {
    const data = await loadNPZ(buffer);

    for (const [key, arr] of Object.entries(data)) {
      switch (key) {
        case 'row_ptrs':
          result.rowPtrs = arr.data as Int32Array;
          break;

        case 'possible_next_states':
          result.possibleNextStatesFlat = arr.data as Int32Array;
          break;

        case 'lp':
          result.transLP = arr.data as Float32Array;
          break;

        case 'states':
          result.statesFlat = arr.data as Int32Array;
          break;

        case 'states_to_idx':
          result.statesToIdxFlat = arr.data as Int32Array;
          result.statesToIdxShape = arr.shape as [number, number, number];
          break;

        case 'period':
          result.initPeriodDist = arr.data as Float32Array;
          break;
      }
    }
  }

  const missing = Object.entries(result).filter(([, v]) => v == null);
  if (missing.length > 0) {
    const names = missing.map(([k]) => k).join(', ');
    throw new Error(`loadArrays(): Missing required fields: ${names}`);
  }

  return result as Arrays;
}

class Node {
  lp: number;
  state: State;
  prev: Node | null;
  frame: number; // need to be rounded when use for indexing

  constructor(lp: number, state: State, prev: Node | null = null, frame = 0) {
    this.lp = lp;
    this.state = state;
    this.prev = prev;
    this.frame = frame;
  }
}

function posPerBeat(period: number, fps: number): number {
  const basePeriod = (60 / 120) * fps; // period for 120 BPM reference
  const basePosPerBeat = 3; // 3 positions per beat at 120 BPM
  return Math.round((basePosPerBeat * period) / basePeriod);
}

function framesPerPosition(period: number, fps: number): number {
  return period / posPerBeat(period, fps);
}

function emissionLP(isDownbeat: boolean, isBeat: boolean, observation: Observation): number {
  const [beatProb, downbeatProb, noBeatProb] = observation;
  const eps = 1e-9;
  if (isDownbeat) return Math.log(downbeatProb + eps);
  if (isBeat) return Math.log(beatProb + eps);
  return Math.log(noBeatProb + eps);
}

function defaultBeamWidth(T: number, c = 450000, wMin = 64, wMax = 1024): number {
  const w = Math.round(c / T);
  return Math.max(wMin, Math.min(wMax, w));
}

// Get observation vector at time index t
function getObservation(nnOutput: Float32Array<ArrayBufferLike>, t: number): Observation {
  // assume 3 channels
  const T = nnOutput.length / 3;
  return [nnOutput[0 * T + t], nnOutput[1 * T + t], nnOutput[2 * T + t]];
}

// Read state at index i from statesFlat
function getState(statesFlat: Int32Array, i: number): State {
  return [statesFlat[i * 3], statesFlat[i * 3 + 1], statesFlat[i * 3 + 2]];
}

function getStateIdx(
  stateToIdxFlat: Int32Array,
  stateToIdxShape: [number, number, number],
  state: State
): number {
  const [periodIdx, tsIdx, phase] = state;
  const [_P, TS, maxPhase] = stateToIdxShape;
  // index formula: ((p * TS) + ts) * maxPhase + phase
  return stateToIdxFlat[(periodIdx * TS + tsIdx) * maxPhase + phase];
}

// Backtrack returns an array of [time_sec, beat_num]
function backtrack(node: Node | null, fps: number, periods: number[]): number[][] {
  const path: Node[] = [];
  let n: Node | null = node;
  while (n !== null) {
    path.push(n);
    n = n.prev;
  }
  path.reverse();

  const beats: number[][] = [];
  let t = 0; // time in frames (but later converted to seconds incrementally)
  for (const nd of path) {
    const [periodIdx, _ts, phase] = nd.state;
    const period = periods[periodIdx];
    if (phase % posPerBeat(period, fps) === 0) {
      const beatNum = Math.floor(phase / posPerBeat(period, fps));
      // append [time_in_seconds, beat_number(1-based)]
      beats.push([t / fps, beatNum + 1]);
    }
    t += framesPerPosition(period, fps);
  }
  return beats;
}

export function beamSearch(
  config: Config,
  nnOutput: Float32Array<ArrayBufferLike>,
  arrays: Arrays,
  progressCallback?: (progress: number) => void,
  beamWidth?: number
): { beats: number[][]; bestLP: number } {
  const T = nnOutput.length / 3;
  const fps: number = config.spectrogram.fps;

  if (beamWidth === undefined || beamWidth === null) {
    beamWidth = defaultBeamWidth(T);
  }

  const periods: number[] = (config.post.tempo_bins as number[])
    .map((tempo) => (60 / tempo) * fps)
    .sort((a, b) => a - b);

  // wrapper for emissionLP
  function calcEmitLP(state: State, observation: Observation): number {
    const periodIdx = state[0];
    const phase = state[2];
    const period = periods[periodIdx];
    const isDownbeat = phase === 0;
    const isBeat = phase % posPerBeat(period, fps) === 0;
    return emissionLP(isDownbeat, isBeat, observation);
  }

  // Unpack arrays
  const rowPtr = arrays.rowPtrs;
  const possibleNextStatesFlat = arrays.possibleNextStatesFlat;
  const transLP = arrays.transLP;
  const statesFlat = arrays.statesFlat;
  const statesToIdxFlat = arrays.statesToIdxFlat;
  const statesToIdxShape = arrays.statesToIdxShape;
  const initPeriodDist = arrays.initPeriodDist;

  const numStates = statesFlat.length / 3;

  // compute start-state probabilities (like python)
  const startStatesProbs = new Float32Array(numStates);
  for (let i = 0; i < numStates; ++i) {
    const st = getState(statesFlat, i);
    const periodIdx = st[0];
    const emitLp = calcEmitLP(st, getObservation(nnOutput, 0));
    startStatesProbs[i] = initPeriodDist[periodIdx] + emitLp;
  }

  // select initial beam (top-k)
  // create an array of indices then partial sort: for simplicity use sorting (numStates ~ 10k ok)
  const idxs = Array.from({ length: numStates }, (_, i) => i);
  idxs.sort((a, b) => startStatesProbs[a] - startStatesProbs[b]); // ascending
  const startStatesIdxs = idxs.slice(Math.max(0, idxs.length - beamWidth));
  let beam: Node[] = startStatesIdxs.map(
    (i) => new Node(startStatesProbs[i], getState(statesFlat, i))
  );

  let bestNode: Node | null = null;
  let bestLP = -Infinity;
  let depth = 0;
  while (beam.length > 0) {
    const node = beam.pop()!;

    const periodIdx = node.state[0];
    const period = periods[periodIdx];
    const nextFrame = node.frame + framesPerPosition(period, fps);

    if (Math.round(nextFrame) >= T) {
      if (node.lp > bestLP) {
        bestNode = node;
        bestLP = node.lp;
      }
      continue;
    }

    depth = Math.max(nextFrame, depth);
    if (progressCallback) progressCallback((depth / T) * 100);

    const nextObservation = getObservation(nnOutput, Math.round(nextFrame));

    const stateIdx = getStateIdx(statesToIdxFlat, statesToIdxShape, node.state);
    const start = rowPtr[stateIdx];
    const end = rowPtr[stateIdx + 1];

    for (let p = start; p < end; ++p) {
      const nextStateIdx = possibleNextStatesFlat[p];

      const nextStateTuple = getState(statesFlat, nextStateIdx);
      const emitLogp = calcEmitLP(nextStateTuple, nextObservation);

      const nextLP = node.lp + transLP[p] + emitLogp;
      beam.push(new Node(nextLP, nextStateTuple, node, nextFrame));
    }

    // keep top beamWidth nodes (same behaviour as python: sort ascending, keep last beamWidth)
    beam.sort((a, b) => a.lp - b.lp);
    if (beam.length > beamWidth) {
      beam = beam.slice(beam.length - beamWidth);
    }
  }

  const beats = backtrack(bestNode, fps, periods);
  return { beats, bestLP };
}
